pkg/kubeconfig/kubeconfig_writer.go

```diff
@@ -0,0 +1,113 @@
+package kubeconfig
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/aws/eks-anywhere/pkg/clients/kubernetes"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/logger"
+	"github.com/aws/eks-anywhere/pkg/retrier"
+)
+
+// ClientFactory builds Kubernetes clients.
+type ClientFactory interface {
+	// BuildClientFromKubeconfig builds a Kubernetes client from a kubeconfig file.
+	BuildClientFromKubeconfig(kubeconfigPath string) (kubernetes.Client, error)
+}
+
+// GenericKubeconfigWriter reads the kubeconfig secret on a cluster and copies the contents to a writer.
+type GenericKubeconfigWriter struct {
+	client  ClientFactory
+	timeout *time.Duration
+	backoff *time.Duration
+}
+
+// WriterOpt allows to configure [KubeconfigWriter].
+type WriterOpt func(*GenericKubeconfigWriter)
+
+// WithTimeout sets the optional timeout for a KubeconfigWriter.
+func WithTimeout(timeout time.Duration) WriterOpt {
+	return func(writer *GenericKubeconfigWriter) {
+		writer.timeout = &timeout
+	}
+}
+
+// WithBackoff sets the optional backoff duration for a KubeconfigWriter.
+func WithBackoff(backoff time.Duration) WriterOpt {
+	return func(writer *GenericKubeconfigWriter) {
+		writer.backoff = &backoff
+	}
+}
+
+func defaultKubeconfigWriterOpts() []WriterOpt {
+	return []WriterOpt{WithTimeout(time.Minute), WithBackoff(time.Second)}
+}
+
+// BuildKubeconfigWriter creates a KubeconfigWriter.
+func BuildKubeconfigWriter(unauthClient ClientFactory, opts ...WriterOpt) GenericKubeconfigWriter {
```

Comment:
for consistency with other constructors
```suggestion
func NewGenericKubeconfigWriter(unauthClient ClientFactory, opts ...WriterOpt) GenericKubeconfigWriter {
```